axlearn/cloud/gcp/measurement.py

```diff
@@ -49,10 +65,50 @@ def record(self, event: measurement.Event, *args, **kwargs):
             self._recorder.record_job_end_time(*args, **kwargs)
         elif event == measurement.Event.START_STEP:
             self._recorder.record_step_start_time(*args, **kwargs)
+        elif event == measurement.Event.START_ACCELERATOR_INIT:
+            self._recorder.record_tpu_init_start_time(*args, **kwargs)
+        elif event == measurement.Event.END_ACCELERATOR_INIT:
+            self._recorder.record_tpu_init_end_time(*args, **kwargs)
+        elif event == measurement.Event.START_TRAINING_PREPARATION:
+            self._recorder.record_training_preparation_start_time(*args, **kwargs)
+        elif event == measurement.Event.END_TRAINING_PREPARATION:
+            self._recorder.record_training_preparation_end_time(*args, **kwargs)
+        elif event == measurement.Event.START_DATA_LOADING:
+            self._recorder.record_data_loading_start_time(*args, **kwargs)
+        elif event == measurement.Event.END_DATA_LOADING:
+            self._recorder.record_data_loading_end_time(*args, **kwargs)
         else:
             logging.log_first_n(
                 logging.WARNING,
                 "Ignoring unknown event %s",
                 1,
                 event,
             )
+
+    def start_monitoring(self, *args, **kwargs):
+        """Starts Monitoring of Goodput.
+
+        Instantiate ml-goodput-measurement's GoodputMonitor to asynchronously calculate
+        Goodput and Badput at the upload_interval and upload to the specified TensorBoard
+        directory.
+        Note: This function requires initialization of distributed JAX before it is called.
+        """
+        if self._monitor is None:
+            cfg: GoodputRecorder.Config = self.config
+            self._monitor = goodput_monitoring.GoodputMonitor(
+                job_name=cfg.name,
+                logger_name=f"goodput_logger_{cfg.name}",
+                tensorboard_dir=cfg.upload_dir,
+                upload_interval=int(cfg.upload_interval),
+                monitoring_enabled=(jax.process_index() == 0),
+                include_badput_breakdown=True,
+            )
+            if not self._monitor:
+                # This could happen if there are internal errors (such as access errors) from GCP services such as Cloud Logging or Cloud Storage.
+                logging.log_first_n(
+                    logging.WARNING,
+                    "Goodput upload could not be started. Please check GoodputMonitor logs.",
+                    1,
+                )
+        self._monitor.start_goodput_uploader(*args, **kwargs)
+        logging.info("Started Goodput upload to Tensorboard in the background!")
```

Comment:
```suggestion
        if self._monitor:
            self._monitor.start_goodput_uploader(*args, **kwargs)
            logging.info("Started Goodput upload to Tensorboard in the background!")
        else:
            # This could happen if there are internal errors (such as access errors) from GCP services such as Cloud Logging or Cloud Storage.
            logging.log_first_n(
                logging.WARNING,
                "Goodput upload could not be started. Please check GoodputMonitor logs.",
                1,
            )
```
So that we check that `self._monitor` is valid before invoking `start_goodput_uploader`.

BTW, it's still unclear to me how `self._monitor` can be None after we construct the instance of `GoodputMonitor`. Are we missing a try/catch somewhere? Does the `__init__` method of `GoodputMonitor` raise an exception (that seems a bit unexpected)?

Reply:
PTAL, I've updated this code. GoodputMonitor's __init__ does not raise exceptions. We check if the monitor is not yet initialized, if so we do that, and then we can safely call start_goodput_uploader. The potential GCP errors we spoke about earlier can only happen after start_goodput_uploader is called. These only log the error and do not have any impact on the workload.